### PR TITLE
Bugfix/translate2

### PR DESCRIPTION
**Pull request by pchlupacek**
This fixes translate bug in M11/RC1 plus
- fixes nested scope interruption (fix by @AdamChlupacek )
- simplifies interruption implementation in Algebra

@SystemFw @mpilquist thanks for comments

### CODE REVIEW

File: core/shared/src/main/scala/fs2/internal/Algebra.scala [hunk 1]
```diff
@@ -341,26 +408,137 @@ private[fs2] object Algebra {
...
+            Algebra.segment(run.values).transformWith { r =>
+              translate0(fK, f(r), effect)
+            }
+
+          case u: Algebra.Uncons[G, x, O] =>
+            uncons[x](u.s, u.chunkSize, u.maxSteps) match {
+              case Done(r) => translate0(fK, f(Right(r)), effect)
+              case Interrupted(_) =>
+                translate0(fK,
+                           f(Left(new Throwable("Translated scope cannot be interrupted"))),
```
Comment (AdamChlupacek) [inline core/shared/src/main/scala/fs2/internal/Algebra.scala]: Seems you forgot to properly implement interruption of translated scope here.

File: core/shared/src/main/scala/fs2/internal/Algebra.scala [hunk 2]
```diff
@@ -63,13 +101,30 @@ private[fs2] object Algebra {
...
+        s.transformWith {
+          case Right(_) => closeScope(scope.id)
+          case Left(err) =>
+            closeScope(scope.id).transformWith {
+              case Right(_)   => raiseError(err)
+              case Left(err0) => raiseError(CompositeFailure(err, err0, Nil))
+            }
+        }
+    }
+  // FreeC.Eval[Algebra[F, O, ?], Unit](OpenScope(s, interruptible))
```
Comment (AdamChlupacek) [inline core/shared/src/main/scala/fs2/internal/Algebra.scala]: Commented out, remove?

File: core/shared/src/main/scala/fs2/internal/Algebra.scala [hunk 3]
```diff
@@ -111,31 +172,39 @@ private[fs2] object Algebra {
...
-    type UnconsF[X] =
-      Function2[
-        CompileScope[F, O],
-        Either[Throwable, Option[(Segment[X, Unit], FreeC[Algebra[F, X, ?], Unit])]],
-        F[(CompileScope[F, O], FreeC[Algebra[F, O, ?], Unit])]
-      ]
+
+    // Uncons is interrupted, fallback on `compileFoldLoop` has to be invoked
+    // The token is a scope from which we should recover.
+    case class Interrupted[X](token: Token) extends UR[X]
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/internal/Algebra.scala]: Can we move these classes to directly under `Algebra` to ensure they don't accidentally close over any state from this method?

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/internal/Algebra.scala]: Oh I guess the close over `F`

Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/internal/Algebra.scala]: Well, there is unfortunatelly disrepance (see Continue) between ones used in compile and translate. In fact closing them inside F/G allowed me to discover another typecast bug. 

Having said that, once we will battle-test this implementation, we could mover them outside compile//translate

**Comment by mpilquist** [comment]
Looks great. Let's merge once you address Adam's comments and then I'll release RC2.

**Comment by pchlupacek** [comment]
ok shall be later today. so we shall be good to go fro RC2 tonite.  Also could you pls test this against scodec before releasing?

**Comment by mpilquist** [comment]
Yep already did and commented on the issue - all tests pass

**Comment by SystemFw** [comment]
Great stuff!

Should be ask for e.g. doobie to give this a try? Or release RC2 directly?
(although I doubt something like scodec won't fail if the implementation was faulty)

**Comment by ChristopherDavenport** [comment]
Testing doobie locally against this branch my only test errors are in postres and examples, which I believe are because I don't have a posgres instance up to test against.
```
[error] (postgres / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] (example / Test / test) sbt.TestsFailedException: Tests unsuccessful
```
I don't know what tests were necessary to check, but all everything not requiring that appears to be passing on the series/0.5.x branch.

**Comment by mpilquist** [comment]
@ChristopherDavenport Thanks very much for testing!

@pchlupacek Ready for merge?

**Comment by ChristopherDavenport** [comment]
I have now gotten the DB operation and 100% of doobie tests now passed.

**Comment by pchlupacek** [comment]
@ChristopherDavenport cool and thanks. @mpilquist yes, we could make a build.

**Pull request merged**